### ISSUE TITLE
'sudo: simple-file-writer: command not found' when resizing 

### ISSUE THREAD

**Issue by Routhinator** [democratic-csi/democratic-csi#390]
This error seems new. It's now coming up when resizing a volume on the new TrueNas Scale Dragonfish release.. is this executable supposed to be on the TrueNAS side?

Driver `freenas-iscsi-csi` - Democratic CSI Chart 0.14.6
TrueNAS Scale Dragonfly Train - Version TrueNAS-SCALE-24.04.0

```
0s          Warning   VolumeResizeFailed         persistentvolumeclaim/nextcloud-db-3          resize volume "pvc-52d505c0-91b1-451f-a168-b66506e14846" by resizer "org.democratic-csi.iscsi" failed: rpc error: code = Unknown desc = error reloading iscsi daemon: {"stderr":"sudo: simple-file-writer: command not found\n","code":1}
```

**Comment by Routhinator** [comment]
Actually, after chasing this down for a but, i believe this is related to #295

**Comment by Routhinator** [comment]
I've rolled back to `v1.9.0` for now, and back to `sudo: simple-file-writer: command not found`

@travisghansen if you have any ideas on how i can get the resize unstuck for now, that would be appreciated.

**Comment by Routhinator** [comment]
Well, the command exists in the image and is in the path, I'm confused why the CSI cannot find it during execution...

```shell
-> % docker run -it --entrypoint bash ghcr.io/democratic-csi/democratic-csi:v1.9.0
root@9ceb6701b729:/home/csi/app# ls
bin  csi_proto  csi_proxy_proto  LICENSE  node_modules  package.json  package-lock.json  src
root@9ceb6701b729:/home/csi/app# find / -type f -name "simple-file-writer"
/usr/local/bin/simple-file-writer
root@9ceb6701b729:/home/csi/app# echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@9ceb6701b729:/home/csi/app# simple-file-writer
/usr/local/bin/simple-file-writer: line 3: ${2}: ambiguous redirect
root@9ceb6701b729:/home/csi/app# cat /usr/local/bin/simple-file-writer
#!/bin/bash

echo ${1} > ${2}
root@9ceb6701b729:/home/csi/app#
```

**Comment by Routhinator** [comment]
I have a feeling this `expandVolume` function somehow doesn't have the $PATH the same as the container, and thus it may not be picking up `/usr/local/bin`  - testing a longshot here, I've jump into the CSI and edit the ssh.js to use the full path instead of just the wrapper script name, to see if that's it.

https://github.com/democratic-csi/democratic-csi/blob/a6dec24a70c393da9d3643c2977b7c57cbe6c2c6/src/driver/freenas/ssh.js#L2057

**Comment by Routhinator** [comment]
Ok, took a few iterations of building an image variant to confirm since I can't build the container due to not being able to download objectivefs but:

```shell
{"host":"truenas-iscsi-democratic-csi-controller-5f4b6cfd5d-vjsrk","level":"error","message":"handler error - driver: FreeNASSshDriver method: ControllerExpandVolume error: {\"name\":\"GrpcError\",\"code\":2,\"message\":\"error reloading iscsi daemon: {\\\"stderr\\\":\\\"sudo: /usr/local/bin/simple-file-writer: command not found\\\\n\\\",\\\"code\\\":1}\"}","service":"democratic-csi","timestamp":"2024-04-26T21:36:25.603Z"}

```

It's not a path problem - same error with the full path. I think this is being executed on the FreeNAS host?


Another issue I hit, while looking at this: 

My cert expired on the TrueNAS and I started getting

```
{"host":"truenas-iscsi-democratic-csi-controller-64bcdd9859-jlh6w","level":"error","message":"handler error - driver: FreeNASSshDriver method: Probe error: TypeError: err.getMessage is not a function TypeError: err.getMessage is not a function\n    at FreeNASSshDriver.Probe (/home/csi/app/src/driver/freenas/ssh.js:49:46)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async requestHandlerProxy (/home/csi/app/bin/democratic-csi:222:18)","service":"democratic-csi","timestamp":"2024-04-26T21:22:46.912Z"}

```

So the error message I posted showing the full path looks weird because to find out the cert expired, I had to modify this:

```javascript
        throw new GrpcError(
          grpc.status.FAILED_PRECONDITION,
          `TrueNAS api is unavailable: ${err.getMessage()}`
        );
```

to 

```javascript
        throw new GrpcError(
          grpc.status.FAILED_PRECONDITION,
          `TrueNAS api is unavailable: ${err}`
        );
```

That allowed me to see the cert error - seems like whatever throws the cert error isn't building the error object the same way.

**Comment by Routhinator** [comment]
So I noticed the TODO and realized the if statement causing this and shell script aren't needed if you double quote the echo statement for the sudo call after playing around in the TrueNAS shell.

```shell
csi@truenas01:~$ sudo echo 1 > /sys/kernel/scst_tgt/devices/csi-pvc-bf920f1b-9270-437c-9193-8724cf1eee24-cluster/resync_size
-bash: /sys/kernel/scst_tgt/devices/csi-pvc-bf920f1b-9270-437c-9193-8724cf1eee24-cluster/resync_size: Permission denied
csi@truenas01:~$ sudo sh -c echo 1 > /sys/kernel/scst_tgt/devices/csi-pvc-bf920f1b-9270-437c-9193-8724cf1eee24-cluster/resync_size
-bash: /sys/kernel/scst_tgt/devices/csi-pvc-bf920f1b-9270-437c-9193-8724cf1eee24-cluster/resync_size: Permission denied
csi@truenas01:~$ sudo sh -c "echo 1 > /sys/kernel/scst_tgt/devices/csi-pvc-bf920f1b-9270-437c-9193-8724cf1eee24-cluster/resync_size"

```

So I modified your original code from 

```javascript
          if (process.env.DEMOCRATIC_CSI_IS_CONTAINER == "true") {
            // use the built-in wrapper script that works with sudo
            command = execClient.buildCommand("simple-file-writer", [
              "1",
              `/sys/kernel/scst_tgt/devices/${kName}/resync_size`,
            ]);
          } else {
            // TODO: syntax fails with sudo
            command = execClient.buildCommand("sh", [
              "-c",
              `echo 1 > /sys/kernel/scst_tgt/devices/${kName}/resync_size`,
            ]);
          }
```

To 

```javascript

          command = execClient.buildCommand("sh", [
            "-c",
            `"echo 1 > /sys/kernel/scst_tgt/devices/${kName}/resync_size"`,
          ]);
```

And that has resolved it for me.

**Comment by Routhinator** [comment]
I'm currently running off my patched version `registry.gitlab.com/routhio/docker/democratic-csi:v1.9.6` - this is public so if you want to see the changes you can pull that.

I can throw up a PR -however I'm uncertain how to properly test the build with the objectivefs part of the build blocking me from building.

**Comment by travisghansen** [comment]
You are entirely correct about running on the TN machine. Not sure what I was thinking. Is the code you have working with sudo?

**Comment by Routhinator** [comment]
Yes

**Comment by rouke-broersma** [comment]
I have similar issues but my errors are different, is this the same issue or should I create a new one?

Kubernetes event:
```
NodeExpandVolume.NodeExpandVolume failed for volume "pvc-3e0e9bd7-5c3a-418f-b6a4-6f008acccca7" : Expander.NodeExpand failed to expand the volume : rpc error: code = Internal desc = {"code":1,"stdout":"Filesystem at /dev/sdl is mounted on /var/lib/kubelet/plugins/kubernetes.io/csi/org.democratic-csi.iscsi.hdd/0ee0287604124898644c00f65443d390c8f67b62f865064d56ec33147efbf033/globalmount; on-line resizing required\nold_desc_blocks = 1, new_desc_blocks = 1\n","stderr":"resize2fs 1.47.0 (5-Feb-2023)\nresize2fs: Permission denied to resize filesystem\n","timeout":false}
```

Dmesg:

```
kern: warning: [2024-04-30T15:18:05.903900173Z]: EXT4-fs warning (device sdl): ext4_resize_begin:83: There are errors in the filesystem, so online resizing is not allowed
```

I am coincidentally also blocked on a postgres pod.

**Comment by Routhinator** [comment]
That is a different error and not related.

And the exec error you get with my image suggests you are not running it on x86 infrastructure. I did not build arm64 images

**Comment by rouke-broersma** [comment]
> That is a different error and not related.
> 
> And the exec error you get with my image suggests you are not running it on x86 infrastructure. I did not build arm64 images

Ah I see, it is indeed only my arm node that has the issue.

**Comment by travisghansen** [comment]
https://github.com/democratic-csi/democratic-csi/issues/295

That syntax was reported to not work previously. Need to figure what the deal is here..

**Comment by travisghansen** [comment]
Oh, I see yours has quotes..nevermind :)

**Comment by travisghansen** [comment]
Should be fixed here: https://github.com/democratic-csi/democratic-csi/commit/38bee217dd4806b27e62ecb3fc4d335e004a9052

Give `v1.9.1` a try.
